### ISSUE TITLE
Support direct creation of HttpMessageHandler instances?

### ISSUE THREAD

**Issue by martincostello** [aspnet/HttpClientFactory#110]
_Related to IdentityServer/IdentityServer4#2311_

A number of libraries provide ways to interact with their internal HTTP calls via allowing the integrator to specify an `HttpMessageHandler` implementation to use. Examples include:
  1. [`RemoteAuthenticationOptions.BackchannelHttpHandler`](https://github.com/aspnet/Security/blob/f8b4f4c620e2faaeec262d302df143068c8d0b33/src/Microsoft.AspNetCore.Authentication/RemoteAuthenticationOptions.cs#L74) used by various `Microsoft.AspNetCore.Authentication.*` packages.
  1. [`IdentityServerAuthenticationOptions.JwtBackChannelHandler`](https://github.com/IdentityServer/IdentityServer4.AccessTokenValidation/blob/3fd88d9ae73734504e269b08901779a7bac1484f/src/IdentityServer4.AccessTokenValidation/IdentityServerAuthenticationOptions.cs#L100) which is part of [`IdentityServer4.AccessTokenValidation`](https://github.com/IdentityServer/IdentityServer4.AccessTokenValidation).

From looking at the [`DefaultHttpClientFactory`](https://github.com/aspnet/HttpClientFactory/blob/9edb6d12d039efdda6a8038227b146bfd12f7ab5/src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs#L117-L118) implementation, it seems like there's currently no way to be able to create an `HttpMessageHandler` directly while retaining the named-configuration, cycling, logging etc. capabilities/behaviours gained when using `IHttpClientFactory` to create `HttpClient` instances.

As a very rough pseudo-implementation, I think what would allow easy interop for application developers to leverage the HttpClientFactory features with existing libraries would be something like this:

```csharp
public interface IHttpMessageHandlerFactory
{
    HttpMessageHandler CreateHandler(string name);
}
```

Which could then be implemented in `DefaultHttpClientFactory` by something like:

```csharp
public HttpMessageHandler CreateHandler(string name)
{
    if (name == null)
    {
        throw new ArgumentNullException(nameof(name));
    }

    var entry = _activeHandlers.GetOrAdd(name, _entryFactory).Value;
    var handler = new LifetimeTrackingHttpMessageHandlerWrapper(entry.Handler);

    StartHandlerEntryTimer(entry);

    return handler;
}

internal class LifetimeTrackingHttpMessageHandlerWrapper : DelegatingHandler
{
    internal LifetimeTrackingHttpMessageHandler(LifetimeTrackingHttpMessageHandler innerHandler)
        : base(innerHandler)
    {
    }

    protected override void Dispose(bool disposing)
    {
        // Do not dispose - lifetime is tracked separately by DefaultHttpClientFactory
    }
}
```

I realise that if this is a valid use case that could be provided, it's probably far too late in the day to ship as part of 2.1.

**Comment by stevejgordon** [comment]
Hey @martincostello. There's a couple of extensions on the IHttpClientBuilder that you can call when defining named/typed clients. If I've understood correctly those might give you the integration points you need.

ConfigurePrimaryHttpMessageHandler lets you pass a Func<HttpMessageHandler>.
ConfigureHttpMessageHandlerBuilder gives you access to the HttpMessageHandlerBuilder.

**Comment by martincostello** [comment]
I’ll take another look shortly, but it’s not configurability I’m lacking, but rather than ability to get a message handler by name (e.g. the GitHub one) to make HTTP calls with without it being wrapped into an HttpClient. I did find the builder in DI, but it seemed to just give me a default `HttpClientHandler` instance, so was missing my custom handler chain.

**Comment by martincostello** [comment]
@stevejgordon So I can access an `HttpMessageHandlerBuilder` through an `IServiceCollection` instance, but that just returns a "vanilla" instance with a null name and with none of the configured additional handlers for logging etc, so the handler returned by `Build()` is just an instance of `HttpClientHandler` which is the value of the builder's `PrimaryHandler` property.

**Comment by martincostello** [comment]
Have raised a PR that could potentially implement this scenario as #118.

**Comment by rynowak** [comment]
Thanks @martincostello - I think this is something we should consider. I've gone on a similar journey to you - but I have question **why** those libraries don't accept an HttpClient themselves.

My gut feeling having looked through some of the Azure-related libraries is that they reason they want a handler rather than a client is because they layer their own opinionated functionality on top of the handler. Many libraries (both from Microsoft and the community) have built up their own layers of policies for things like auth and retries over time. This is just entropy 😆 users will ask for features and library authors will deliver them.

One of my favourite things about the factory is that the developer gets to be in control and build a rich or complicated pipeline if the requirements dictate it, and what you see is what you get. 

----

We're currently going through an investigation with the Azure SDK team - and the question on my mind is whether a library like the Azure Storage SDK would want to have separate code paths like:
- user gave me a handler - layer my own opinions on top
- user gave me a client - trust that it's configured in a reasonable way

This is a general architectural problem for .NET libraries that do HTTP that we need to figure out (with consensus from the community).

You could help by providing some examples - are there libraries that treat an HttpClient and a handler identically? Are there libraries that accept a handler but **don't** layer additional policies on top?

Basically what I want to settle is: what would be our guidance for a library that uses HTTP internally (examples: Octokit, Azure SDK, Identity Service, etc). 

---

If the ask here is **just related to the connection and lifetime management features** of the factory then I think we'd say no. The right fix for .NET is for HttpClient to have a reasonable behaviour out of the box. You shouldn't need an abstraction for that.

**Comment by rynowak** [comment]
I'm tagging this to at least consider. We should starting working on this problem even if we don't do what was proposed here

**Comment by martincostello** [comment]
Thanks for the detailed reply @rynowak. In my specific case, what I was trying to achieve was the ability get into the HTTP pipeline for test purposes using [this library](https://github.com/justeat/httpclient-interception) so I can do "in-memory" integration tests of things, not to do _connectiony stuff_. Effectively, I wanted to leverage the handler-building features.

For a specific example of the kind of thing I was doing, you can check out this PR into a side project of mine (though it's a bit long as it snowballed...): martincostello/alexa-london-travel-site#186. I was ultimately successful in using HttpClientFactory to setup integration tests for my 3rd-party auth integrations, greatly improving my test coverage for something that would otherwise have too many moving parts to be able to regularly test with automation.

My assumption is that libraries offered `HttpMessageHandler` as the injection point as it was the "mockable" part, probably for the purposes of their own tests, or as an extension point for delegating handlers. As you mention, the intentions for `HttpMessageHandler` vs. `HttpClient` by library authors aren't always clearly documented.

As you mention Azure in your reply, I actually raised an issue against the CosmosDB SDK a while ago for the same reasons, as their `HttpClient` usage is heavily hidden deep in the library from what I've gleaned with a decompiler. I wanted to be able to do the same there to do integration tests against CosmosDB (or at least have the option to, depending on complexity), as provisioning the emulator Docker image in a CI scenario was far too slow, and I didn't want to have to maintain and pay for a whole instance just for testing in a side project: Azure/azure-documentdb-dotnet#245. It would also have enabled me to add custom ApplicationInsights telemetry to collect HTTP response headers of resource usage, which eventually evolved into this issue: Microsoft/ApplicationInsights-dotnet-server#587.

To sum up, three examples off the top of my head are:
  1. CosmosDB - `HttpClient` internally as a block-box, no configuration possible.
  1. Microsoft Identity Middleware - Provides both, but if you provide an `HttpClient`, you don't get some defaults they apply to it ([example](https://github.com/aspnet/Security/blob/b3e4bf382c9676e2d912636b7ee0255cad244e11/src/Microsoft.AspNetCore.Authentication.Twitter/TwitterPostConfigureOptions.cs#L42-L47)). Otherwise, it's wrapped into an `HttpClient` they new-up directly.
  1. IdentityServer - Provide an `HttpMessageHandler`, which they just use directly if provided, otherwise they make an `HttpClientHandler`, and then [new-up an `HttpClient`](https://github.com/IdentityServer/IdentityServer4.AccessTokenValidation/blob/3fd88d9ae73734504e269b08901779a7bac1484f/src/IdentityServer4.AccessTokenValidation/IdentityServerAuthenticationOptions.cs#L165-L169).

**Comment by rynowak** [comment]
Thanks, that's a really great set of info. I can't promise any immediate action on this, but I think we need to do *something* to improve this. That's a great use case for a feature like this.

That might mean something like your PR, or docs and blogs, or improvements to other libraries, or all of the above.

**Comment by martincostello** [comment]
You should be able to set your own certificate validation logic by using `ConfigurePrimaryMessageHandler()` on the `IHttpClientBuilder` retuned by the `AddHttpClient()` methods.

**Comment by bloomtom** [comment]
I considered that, but the certificate hash isn't known on startup. Only the controller has absolute knowledge of this information as it's user provided.

There's also the matter of knowing that a request failed inside my custom validation logic. Without that, I can only say generically "There was a certificate issue", not "Your pin doesn't match the certificate". I can think of some really dirty solution for that (separately maintain and inject some kind of error log), but nothing I'd actually write.

**Comment by martincostello** [comment]
Ah I see. All the things I can think off of the top of my head involve stuffing some context of some sort with your pins/hashes into the Properties on an HttpRequestMessage and then pulling them out of the request on the validation callback, but that’d be quite hacky.

**Comment by bloomtom** [comment]
Indeed. There are "solutions". All of them feel quite radioactive.
Thanks for the help.

**Comment by glatzert** [comment]
I'd also vote for the ability to access the message handler more easily.
We needed "UseDefaultCredentials" and ended up with:

`services.Configure<HttpClientFactoryOptions>(opt => opt.HttpMessageHandlerBuilderActions.Add(builder => ((HttpClientHandler)builder.PrimaryHandler).UseDefaultCredentials = true));`

Which works, but is kind of non-intuitive.

**Comment by rynowak** [comment]
@glatzert you can do this with `ConfigurePrimaryHttpMessageHandler`

```
services.AddHttpClient("foo").ConfigurePrimaryHttpMessageHandler(new HttpClientHandler() { UseDefaultCredentials = true })
```

**Comment by dgwaldo** [comment]
We too have found a use case where we'd like to create HttpClientHandlers at run time and pass them off to the HttpClientFactory. We have a middleware proxy layer that is configurable and needs to create httpClients at run time based on the URL being requested and the credentials which come from configuration. With IHttpClientFactory currently all we can do is hack our way around it by doing something as suggested by @martincostello 4 messages above...

**Comment by rynowak** [comment]
@dgwaldo - can you tell me a little more about your requirements? I'm not totally certain that what we've been discussing here will help you. Maybe I'm missing something though 😕

**Comment by dgwaldo** [comment]
@rynowak  I want to use HttpClientFactory to gain all the benefits of logging etc. I have a proxy that used to just create HttpClients on the fly request as needed. The credentials could then be set at run-time via HttpMessageHandler. Now I'm trying to work around the issue of not being able set these at run-time by spinning through the config at startup... Though this isn't working yet.
`

	var serviceProvider = services.BuildServiceProvider();
			var evilEvilEsriProxyConfig = serviceProvider.GetService<IEvilEsriProxyConfigService>();
			evilEvilEsriProxyConfig.Config.ServerUrls.ToList().ForEach(su => {
				services.AddHttpClient(su.Domain)
					.ConfigurePrimaryHttpMessageHandler(h => {
						return new HttpClientHandler {
							AllowAutoRedirect = false,
							Credentials = evilEvilEsriProxyConfig.GetCredentials(evilEvilEsriProxyConfig.GetProxyServerUrlConfig((su.Url)))
						};
					});
			});
`

**Comment by ilya-chumakov** [comment]
Thanks for your work! I hope to use `HttpClientFactory` one day but I'd like to ask to add a factory method for handlers. We have a scenario similar to mentioned before, when one of the handlers should be configured at runtime, so it's impossible to resolve it by DI. For example: 

    public HttpClient CreateClientAsync(string userId)
    {
        //I'd prefer to use a factory method instead
        var innerHandler = new HttpClientHandler(); 

        var outerHandler = new OuterHandler(userId, innerHandler); 
        
        return new HttpClient(outerHandler);
    }
    
    
P.S. is it a good idea to fork the library and try to use [`CreateHandlerEntry`](https://github.com/aspnet/HttpClientFactory/blob/master/src/Microsoft.Extensions.Http/DefaultHttpClientFactory.cs#L132) method to create `innerHandler`?

**Comment by dgwaldo** [comment]
@rynowak  I think that would work, but if we are swimming upstream here maybe it doesn't make sense? I'd rather keep my work around and have the lifetime management... Maybe this is just something that workarounds could be shown for in the documentation?

**Comment by martincostello** [comment]
Lifetime tracking or not aside, the problem I'm trying to solve is the ability to avoid having a hack like this in my code so I can use `HttpClientFactory` with existing code that assumes that `HttpMessageHandler` is the level of abstraction:

```csharp
var factory = serviceProvider.GetRequiredService<IHttpClientFactory>();
HttpClient client = factory.CreateClient("MyClient");

FieldInfo field = typeof(HttpMessageInvoker).GetField(
    "_handler",
    BindingFlags.GetField | BindingFlags.Instance | BindingFlags.NonPublic);

options.JwtBackChannelHandler = field.GetValue(client) as HttpMessageHandler;
```

**Comment by shaynevanasperen** [comment]
What I'd really like is to be able to create my own class which derives from `HttpClient` so that all the existing methods are usable directly without me needing to wrap them all. Something like this:
```cs
public class JsonPlaceHolderHttpClient : HttpClient
{
    public JsonPlaceHolderHttpClient(HttpMessageHandler innerHandler)
        : base(new EnsureSuccessHandler(innerHandler))
    {
        BaseAddress = new Uri("https://jsonplaceholder.typicode.com")
    }
}

public class EnsureSuccessHandler : DelegatingHandler
{
    public EnsureSuccessHandler(HttpMessageHandler innerHandler)
        : base(innerHandler) { }

    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var response = await base.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        return response;
    }
}
```
And then just do the normal:
```cs
services.AddHttpClient<JsonPlaceHolderHttpClient>();
```
But instead of that, I have to do this ugly hack:
```cs
public class JsonPlaceHolderHttpClient : HttpClient
{
    public JsonPlaceHolderHttpClient(HttpClient httpClient)
        : base(new EnsureSuccessHandler(httpClient.GetHandler()))
    {
        BaseAddress = new Uri("https://jsonplaceholder.typicode.com")
    }
}

public static class HttpClientExtensions
{
    static readonly FieldInfo HandlerFieldInfo = typeof(HttpMessageInvoker)
        .GetField("_handler", BindingFlags.NonPublic | BindingFlags.Instance);

    public static HttpMessageHandler GetHandler(this HttpClient httpClient) =>
        (HttpMessageHandler)HandlerFieldInfo.GetValue(httpClient);
}
```

**Comment by shaynevanasperen** [comment]
Upon further investigation, it turns out that the reason I was getting an `ObjectDisposedException` was because I was trying to resolve an instance of my `JsonPlaceHolderHttpClient` from within a scoped `IServiceProvider`. If I changed the IoC registration of the root of the dependency graph to be a singleton, it worked every time, however I can't do that as I need the root to be a scoped dependency. After some futher playing around, I came up with this slight modification of my hack, which now works properly:

```cs
public class JsonPlaceHolderHttpClient : HttpClient
{
    public JsonPlaceHolderHttpClient(IHttpClientFactory httpClientFactory)
        : base(new EnsureSuccessHandler(httpClientFactory.CreateClient().GetHandler()))
    {
        BaseAddress = new Uri("https://jsonplaceholder.typicode.com");
    }
}

public class EnsureSuccessHandler : DelegatingHandler
{
    public EnsureSuccessHandler(HttpMessageHandler innerHandler)
        : base(innerHandler) { }

    protected override async Task<HttpResponseMessage> SendAsync(
        HttpRequestMessage request, CancellationToken cancellationToken)
    {
        var response = await base.SendAsync(request, cancellationToken);
        response.EnsureSuccessStatusCode();
        return response;
    }
}

public static class HttpClientExtensions
{
    static readonly FieldInfo HandlerFieldInfo = typeof(HttpMessageInvoker)
        .GetField("_handler", BindingFlags.NonPublic | BindingFlags.Instance);

    public static HttpMessageHandler GetHandler(this HttpClient httpClient) =>
        (HttpMessageHandler)HandlerFieldInfo.GetValue(httpClient);
}
```
And then I just do:
```cs
services.AddHttpClient();
services.AddSingleton<JsonPlaceHolderHttpClient>();
```
The difference is that I now only register the standard (non-typed) `IHttpClientFactory` in startup, and then additionally register a singleton of my client class which derives from `HttpClient` but this time it takes an instance of `IHttpClientFactory` in it's constructor and then uses reflection to get at the handler inside it, which I then wrap with my own `DelegatingHandler` and pass to the base constructor.

**Comment by rynowak** [comment]
Closing this since we provided a new API to create the handlers directly.
